client/my-sites/site-settings/settings-p2/preapproved-domains.jsx

```diff
@@ -0,0 +1,264 @@
+import { Card } from '@automattic/components';
+import { ToggleControl } from '@wordpress/components';
+import { useState, useEffect } from '@wordpress/element';
+import classNames from 'classnames';
+import { flowRight, includes, pickBy, filter } from 'lodash';
+import { connect } from 'react-redux';
+import QuerySiteSettings from 'calypso/components/data/query-site-settings';
+import FormFieldset from 'calypso/components/forms/form-fieldset';
+import FormLabel from 'calypso/components/forms/form-label';
+import FormSettingExplanation from 'calypso/components/forms/form-setting-explanation';
+import TokenField from 'calypso/components/token-field';
+import wpcom from 'calypso/lib/wp';
+import SettingsSectionHeader from 'calypso/my-sites/site-settings/settings-section-header';
+import isSiteP2Hub from 'calypso/state/selectors/is-site-p2-hub';
+import isSiteWPForTeams from 'calypso/state/selectors/is-site-wpforteams';
+import { getSelectedSiteId } from 'calypso/state/ui/selectors';
+import wrapSettingsForm from '../wrap-settings-form';
+
+const P2PreapprovedDomainsForm = ( props ) => {
+	const SETTING_KEY_PREAPPROVED_DOMAINS = 'p2_preapproved_domains';
+
+	const {
+		fields,
+		handleSubmitForm,
+		isP2HubSite,
+		isRequestingSettings,
+		isSavingSettings,
+		isWPForTeamsSite,
+		recordTracksEvent,
+		site,
+		translate,
+		updateFields,
+	} = props;
+
+	const [ isToggledOn, setIsToggledOn ] = useState( false );
+	const [ isValidating, setIsValidating ] = useState( false );
+	const [ validTokens, setValidTokens ] = useState( [] );
+	const [ invalidTokens, setInvalidTokens ] = useState( {} );
+	const [ error, setError ] = useState( {} );
+
+	useEffect( () => {
+		if ( ! fields || ! fields.p2_preapproved_domains ) {
```

Comment:
Perhaps we could use the optional chaining like `!fields?.p2_preapproved_domains`?

Reply:
Done!